sparta/src/StatisticInstance.cpp

```diff
@@ -0,0 +1,770 @@
+                      "that was recreated from a SimDB record");
+
+        start_tick_ = getScheduler_()->getElapsedTicks();
+        end_tick_ = Scheduler::INDEFINITE;
+
+        if(user_calculated_si_value_){
+            initial_.resetValue(user_calculated_si_value_->getCurrentValue());
+            result_ = NAN;
+            return;
+        }
+
+        if(sdef_ != nullptr){
+            if(node_ref_.expired() == true){
+                throw SpartaException("Cannot start() a StatisticInstance referring to a "
+                                      "destructed StatisticDef");
+            }
+            stat_expr_.start();
+            initial_.resetValue(0);
+        }else if(ctr_){
+            if(node_ref_.expired() == true){
+                throw SpartaException("Cannot start() a StatisticInstance referring to a "
+                                      "destructed Counter");
+            }
+            initial_.resetValue(ctr_->get());
+        }else if(par_){
+            if(node_ref_.expired() == true){
+                throw SpartaException("Cannot start() a StatisticInstance referring to a "
+                                      "destructed Parameter");
+            }
+            initial_.resetValue(par_->getDoubleValue());
+        }else{
+            stat_expr_.start();
+        }
+
+        // Clear result value
+        result_ = NAN;
+    }
+
+    void StatisticInstance::end(){
+        sparta_assert(direct_lookup_si_value_ == nullptr,
+                      "You cannot call StatisticInstance::end() for an SI "
+                      "that was recreated from a SimDB record");
+
+        end_tick_ = getScheduler_()->getElapsedTicks();
+
+        if(sdef_ != nullptr){
+            if(node_ref_.expired() == true){
+                throw SpartaException("Cannot end() a StatisticInstance referring to a "
+                                      "destructed StatisticDef");
+            }
+            stat_expr_.end();
+        }else if(ctr_ != nullptr){
+            if(node_ref_.expired() == true){
+                throw SpartaException("Cannot end() a StatisticInstance referring to a "
+                                      "destructed Counter");
+            }
+            // Do nothing to counter
+        }else if(par_ != nullptr){
+            if(node_ref_.expired() == true){
+                throw SpartaException("Cannot end() a StatisticInstance referring to a "
+                                      "destructed Parameter");
+            }
+            // Do nothing to Parameter
+        }else{
+            stat_expr_.end();
+        }
+
+        // Recompute result value
+        result_ = computeValue_();
+    }
+
+    double StatisticInstance::getValue() const {
+        if (direct_lookup_si_value_ != nullptr) {
+            return computeValue_();
+        }
+
+        if(end_tick_ < start_tick_){
+            throw ReversedStatisticRange("Range is reversed. End < start");
+        }
```

Comment:
This seems like something that could be checked when `end_tick_` and `start_tick_` are assigned, unless those change frequently and the `getValue` is called infrequently.